04-Protocol-Security.md

```diff
@@ -145,8 +153,11 @@ The following functions will also be referenced:
   - Calls `MixHash(ciphertext)`
   - returns `plaintext`
 
-- `ECDH(k, rk)`: performs an Elliptic-Curve Diffie-Hellman operation using `k`, which is a valid `secp256k1` private key, and `rk`, which is a valid public key
-  - The output is X-coordinate of the resulting EC point
+- `ellswift_ecdh_xonly(k, rk)`: performs an Elliptic-Curve Diffie-Hellman operation using `k`, which is a valid `secp256k1` private key, and `rk`, which is a EllSwift
+encoded public key
+  - The output is 32 bytes
```

Comment:
It might be worth mentioning in more detail what this actually does.

the BIP 324 paper describes this equally named function this way:

* *ellswift_ecdh_xonly(ellswift_theirs, priv)*:
    * Let *u = int(ellswift_theirs[:32]) mod p*
    * Return *bytes(x(priv⋅lift_x(XSwiftEC(u, t))))*
  
It suggest that the output is 32-byte scalar.

Rust binding to secp256k1 however exposes interface probably to something different, because it asks you for a hash-fn (and defaults to sha256)

Reply:
From the secp256k1 binding for rust:

```rust
    /// Computes a shared secret, just like `shared_secret`, but with a custom hash function
    /// for computing the shared secret. For compatibility with other libraries, you should
    /// use `shared_secret` instead, which is already compatible with BIP324.
    /// The hash function takes three arguments: the shared point, and the `ElligatorSwift`
    /// encodings of the two parties and returns a 32-byte shared secret.
    pub fn shared_secret_with_hasher<F>(
        ellswift_a: ElligatorSwift,
        ellswift_b: ElligatorSwift,
        secret_key: SecretKey,
        party: ElligatorSwiftParty,
        mut hash_function: F,
    ) -> ElligatorSwiftSharedSecret
    where
        F: FnMut([u8; 32], [u8; 64], [u8; 64]) -> ElligatorSwiftSharedSecret,
    {
```

My understanding is that I should use this method and define my own function that takes just the first argument and outputs it directly as ElligatorSwiftSharedSecret.
That secret is then processed with noise-prescribed key-derivation functions (HKDF)

Reply:
Is that what https://github.com/stratum-mining/stratum/pull/737 is doing? All I know is that PR is compatible with my Bitcoin Core implementation, and I mostly copy-pasted the EllSwift calls from its BIP324 implementation :-)

Reply:
This is exactly what stratum-mining/stratum#737

I think that PR implements this algorithm:
```
def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
    ecdh_point_x32 = ellswift_ecdh_xonly(ellswift_theirs, priv)
    if initiating:
        # Initiating, place our public key encoding first.
        return sha256_tagged("bip324_ellswift_xonly_ecdh", ellswift_ours + ellswift_theirs + ecdh_point_x32)
    else:
        # Responding, place their public key encoding first.
        return sha256_tagged("bip324_ellswift_xonly_ecdh", ellswift_theirs + ellswift_ours + ecdh_point_x32)
```

Reply:
I don't think the PR stratum-mining/stratum#737 is using algorithm described in this current proposal.

I think it uses this algorithm (see https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki#shared-secret-computation):
```
def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
    ecdh_point_x32 = ellswift_ecdh_xonly(ellswift_theirs, priv)
    if initiating:
        # Initiating, place our public key encoding first.
        return sha256_tagged("bip324_ellswift_xonly_ecdh", ellswift_ours + ellswift_theirs + ecdh_point_x32)
    else:
        # Responding, place their public key encoding first.
        return sha256_tagged("bip324_ellswift_xonly_ecdh", ellswift_theirs + ellswift_ours + ecdh_point_x32)
```

See line 65 in this PR. It references this exact paragraph but references the ellswift_ecdh_xonly. That's my source of confusion.

Reply:
the v2_ecdh from bip324 document is meant to produce the final shared-secret used for symmetric ciphers.

Since noise-framework provides key-derivation and only requires ECDH operation,
it's not necessary to perform this *wrapped* ecdh in form of v2_ecdh.
It would be enough to use the `ellswift_ecdh_xonly`, as this proposal actually suggests.

Reply:
On the Bitcoin Core side I'm using `ECDHSecret CKey::ComputeBIP324ECDHSecret`: https://github.com/bitcoin/bitcoin/blob/master/src/key.cpp#L352-L370

This sets `secp256k1_ellswift_xdh_hash_function_bip324`, and so it returns a hash of the x-coordinate, not the coordinate itself. I assume that's what you mean by "wrapped"?

The `secp256k1_ellswift_xdh` function in libsecp can be called without `hashfp`. However in order to use that I would have to introduce a new helper function in `CKey`. It would be similar to the `UnhashedECDH` used in Silent Payments:

https://github.com/bitcoin/bitcoin/pull/28122/commits/62fda2c0835ca14e446bc0fdddcc36f18902ff87#diff-6c397f7958e890c05ec03255d9e65943bd8a86e8da39f287ab0ecc2bf6a4bc72R138-R142

Such a function is unsafe to use outside of the Noise protocol context. Silent Payments can't work without this, it really needs the unhashed result. But with Stratum v2 the hash is harmless even if not strictly needed.

So my preference would be to keep the implementation, but change the proposal to use the hash function.

Reply:
Section 4.1 of the Noise paper says about `DH(key_pair, public_key):`:

> Performs a Diffie-Hellman calculation be- tween the private key in key_pair and the public_key and returns an output sequence of bytes of length DHLEN.

In other words, it does not _require_ hashing the result, but doesn't mind either.

> Implementations must handle invalid public keys either by returning some output which is purely a function of the public key and does not depend on the private key, or

This requirement is easiest to met by hashing the x coordinate. Though I'm not sure if any x-coordinate can be an invalid public key, so maybe this problem doesn't exist.

Reply:
I just noticed that `bip324_ellswift_xonly_ecdh` is not just a straight forward hash, it's a tagged hash.

The rationale for tagged hashes is here:
https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#specification

> Cryptographic hash functions are used for multiple purposes in the specification below and in Bitcoin in general. To make sure hashes used in one context can't be reinterpreted in another one, hash functions can be tweaked with a context-dependent tag name, in such a way that collisions across contexts can be assumed to be infeasible

So this could prevent issues where someone reuses an authority key as a key in their wallet. Though that seems extremely unlikely. The tag we're using is BIP324 specific. We could pick a fresh one, but that introduces the complexity of having to use lower level functions. Using the same tag shouldn't cause any clashes since we also hash the `Noise_NX_EllSwiftXonly_ChaChaPoly_SHA256` protocol name.

Reply:
Interesting.

Ok, after spending some time with it. I also agree that the "wrapped" ecdh should be fine.
By "wrapped" I mean the `ellswift_ecdh_xonly` inside of `v2_ecdh` - the tagged-hash of the DH-product alongside the elligators.

In fact, using the result of `ellswift_ecdh_xonly` has slightly weaker security properties, as it allows some transformations of the elligators on the wire (like changing their pubkey parities etc). I don't really want to analyse such cases, as I'm no cryptographer and security expert. So let's stick to the standard methods provided by public libraries.
Thanks for the discussion ;-)